src/db/spanner/models.rs

```diff
@@ -0,0 +1,1292 @@
+    type Target = SpannerDbInner;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+macro_rules! batch_db_method {
+    ($name:ident, $batch_name:ident, $type:ident) => {
+        pub fn $name(&self, params: params::$type) -> Result<results::$type> {
+            batch::$batch_name(self, params)
+        }
+    }
+}
+
+impl SpannerDb {
+    pub fn new(
+        conn: Conn,
+        thread_pool: Arc<::tokio_threadpool::ThreadPool>,
+        coll_cache: Arc<CollectionCache>,
+    ) -> Self {
+        let inner = SpannerDbInner {
+            conn,
+            thread_pool,
+            session: RefCell::new(Default::default()),
+        };
+        SpannerDb {
+            inner: Arc::new(inner),
+            coll_cache,
+        }
+    }
+
+    pub(super) fn get_collection_id(&self, name: &str) -> Result<i32> {
+        if let Some(id) = self.coll_cache.get_id(name)? {
+            return Ok(id);
+        }
+        let spanner = &self.conn;
+
+        let mut sql = ExecuteSqlRequest::default();
+        sql.sql = Some("SELECT id FROM collections WHERE name = @name;".to_string());
+        let mut params = HashMap::new();
+        params.insert("name".to_string(), name.to_string());
+        sql.params = Some(params);
+
+        let session = spanner.session.name.as_ref().unwrap();
+        let result = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        let rv = match result {
+            Err(e) => {
+                println!("err {}", e);
+                // TODO Return error
+                Ok(0)
+            }
+            Ok(res) => {
+                println!("ok {:?}", res.1);
+                match res.1.rows {
+                    Some(row) => Ok(row[0][0].parse().unwrap()),
+                    None => Ok(0),
+                }
+            }
+        };
+        if let Ok(val) = rv {
+            self.coll_cache.put(val, name.to_owned())?;
+        };
+        rv
+        //let id = sql_query("SELECT id FROM collections WHERE name = ?")
+        //    .bind::<Text, _>(name)
+        //    .get_result::<IdResult>(&self.conn)
+        //    .optional()?
+        //    .ok_or(DbErrorKind::CollectionNotFound)?
+        //    .id;
+        //self.coll_cache.put(id, name.to_owned())?;
+    }
+
+    pub(super) fn create_collection(&self, name: &str) -> Result<i32> {
+        // XXX: handle concurrent attempts at inserts
+        let spanner = &self.conn;
+        let session = spanner.session.name.as_ref().unwrap();
+
+        let mut sql = ExecuteSqlRequest::default();
+        sql.sql = Some("SELECT MAX(collectionid) from collections".to_string());
+        let result = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        let id = match result {
+            Err(e) => {
+                println!("err {}", e);
```

Comment:
Might want to make these `dbg!` instead to prevent the logs from filling up.

Reply:
All the error handling needs to be combed over once the tests are running against this code. I'll switch to debug when that happens.